Point Displacement label distance factor

HenrikSpa opened the pull request:
## Description

Adds label distance factor spinbox to Point Displacement symbology
window.

Replaces the hard coded distance "symbolDiagonal/2.0" with
"symbolDiagonal*mLabelDistanceFactor", where mLabelDistanceFactor is set
to the spinbox value.

 It's an old feature request (Point displacement labels adjustable
offsets "https://issues.qgis.org/issues/5945").

Here are images that shows the label placement using different values for the distance:

Ring, label distance factor 0.50, which is equivalent to the previous hardcoded 2.0 value and is also the new default value:
![image](https://user-images.githubusercontent.com/17615137/52196597-b75df080-285c-11e9-94cf-eb2f832afa26.png)

Ring 0.25
![image](https://user-images.githubusercontent.com/17615137/52196690-0b68d500-285d-11e9-9054-626bf01d6761.png)

Ring -0.20. Can be useful depending on the symbol style
![image](https://user-images.githubusercontent.com/17615137/52196728-32270b80-285d-11e9-9cad-12fb3eab4a57.png)

Concentric ring, 0.25
![image](https://user-images.githubusercontent.com/17615137/52196770-57b41500-285d-11e9-9f88-82c62dacbdb9.png)

// Grid has changed. See screenshots in later comments.
~~Grid 0.25
![image](https://user-images.githubusercontent.com/17615137/52196791-6e5a6c00-285d-11e9-8790-9d482779ba63.png)~~

Ring 0.50 with different sized symbols. No change in behaviour compared to previous implementation. The label distance seem to adapt to the largest symbol diagonal (variable symbolDiagonal).
![image](https://user-images.githubusercontent.com/17615137/52196970-2be55f00-285e-11e9-9482-c571bf75cfb2.png)


## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR) // (There are 21 failed tests, but those exist in the build from the current unchanged master also).
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

Commit:
[FEATURE] [needs-docs] Point Displacement label distance factor

Adds label distance factor spinbox to Point Displacement symbology
window. Calculates label distance based on the current symbology
diagonal instead of a general maximum diagonal.

Replaces the hard coded distance "symbolDiagonal/2.0" with
"symbolDiagonal*mLabelDistanceFactor", where mLabelDistanceFactor is set
to the spinbox value.

 It's an old feature request (Point displacement labels adjustable
offsets "https://issues.qgis.org/issues/5945").

HenrikSpa commented:
I'm currently checking if it's easy to have it adapt to each individual symbol size instead of the largest symbol.

HenrikSpa commented:
I've made adjustments to have the label distance depend on the current symbol diagonal instead of maximum diagonal. I will update the pull request.

HenrikSpa commented:
Screenshots of labeldistace depending on symbol size:

Ring, 0.35:
![image](https://user-images.githubusercontent.com/17615137/52209334-e3409c80-2883-11e9-8a47-055f124f62f2.png)

Concentric rings, 0.35:
![image](https://user-images.githubusercontent.com/17615137/52209394-0b300000-2884-11e9-9ae7-aee7604e60e4.png)

Grid, 0.2:
![image](https://user-images.githubusercontent.com/17615137/52209434-2f8bdc80-2884-11e9-8d29-340e19284837.png)

HenrikSpa commented:
I'm adjusting the distance for single features.

HenrikSpa commented:
Added new getters and setters labelDistanceFactor and setLabelDistanceFactor to python unittests.

HenrikSpa commented:
I'm fixing a "bug" for grid labels. The bug is an error in how the label position is calculated while using the grid which causes the labels to be placed way out from the grid.

QGIS version 3.4.4:
![image](https://user-images.githubusercontent.com/17615137/52230014-9d50fc00-28b6-11e9-9700-af438e39ccb1.png)

shahzadlone commented on src/core/symbology/qgspointdisplacementrenderer.cpp:
```diff
@@ -243,8 +256,9 @@ void QgsPointDisplacementRenderer::calculateSymbolAndLabelPositions( QgsSymbolRe
   }
   else if ( nPosition == 1 ) //If there is only one feature, draw it exactly at the center position
   {
+    double side = sqrt( pow( symbolDiagonal, 2 ) / 2.0 );
```
Perhaps ```const``` ?

HenrikSpa replied:
Are you suggesting the side variable to be const or symbolDiagonal should be const?

side is only used once for each single symbol, but it can have different values for each single symbol. symbolDiagonal is the maximum diagonal for each group (and when it's a single symbol its equal to the single symbol diagonal). I guess the side could be const for a single symbol, but it's only used in one place for single symbol so I'm not sure what the benefit would be.

shahzadlone replied:
just better practice i guess? so someone doesn't accidentally modify it

HenrikSpa commented:
I've bugfixed the grid labels (will soon push it):

Same symbol size, factor 0,3
![image](https://user-images.githubusercontent.com/17615137/52264115-29a00500-2931-11e9-833e-5f8147feab70.png)

Same symbol size, factor 0,5
![image](https://user-images.githubusercontent.com/17615137/52264167-49cfc400-2931-11e9-9a4e-3ed9ba3f5995.png)

Different sized symbols, factor 0,3
![image](https://user-images.githubusercontent.com/17615137/52264228-679d2900-2931-11e9-8961-a768f46127b0.png)

Grid size increased by 5 mm
![image](https://user-images.githubusercontent.com/17615137/52264270-800d4380-2931-11e9-8312-2fb3cb806c5d.png)

No grid size increase, more symbols, factor 0,3
![image](https://user-images.githubusercontent.com/17615137/52264340-adf28800-2931-11e9-95f2-7646ffb59863.png)

Factor -0,2
![image](https://user-images.githubusercontent.com/17615137/52264444-eb571580-2931-11e9-998a-7c369dfe8464.png)

Commit:
Fixes labels for point displacement grid.

HenrikSpa commented:
I've currently no plans to change anything more in this pull request. Labels for ring, concentric rings and grid all behave as I want them too (without going for full label engine integration).

nyalldawson commented:
Not stale

nyalldawson commented on src/core/symbology/qgspointdisplacementrenderer.cpp:
```diff
@@ -60,12 +61,21 @@ void QgsPointDisplacementRenderer::drawGroup( QPointF centerPoint, QgsRenderCont
 
   //calculate max diagonal size from all symbols in group
   double diagonal = 0;
+  QList<double> diagonals;
```
QVector would be more appropriate here (faster)

nyalldawson commented on src/core/symbology/qgspointdisplacementrenderer.cpp:
```diff
@@ -242,8 +254,9 @@ void QgsPointDisplacementRenderer::calculateSymbolAndLabelPositions( QgsSymbolRe
   }
   else if ( nPosition == 1 ) //If there is only one feature, draw it exactly at the center position
   {
+    const double side = sqrt( pow( symbolDiagonal, 2 ) / 2.0 );
```
```suggestion
    const double side = std::sqrt( std::pow( symbolDiagonal, 2 ) / 2.0 );
```

nyalldawson commented on src/core/symbology/qgspointdisplacementrenderer.cpp:
```diff
@@ -319,16 +337,42 @@ void QgsPointDisplacementRenderer::calculateSymbolAndLabelPositions( QgsSymbolRe
+          yFactor = -1;
+        }
+
+        side = sqrt( pow( diagonals.at( symbolIndex ), 2 ) / 2.0 );
```
```suggestion
        side = std::sqrt( std::pow( diagonals.at( symbolIndex ), 2 ) / 2.0 );
```

nyalldawson commented on src/core/symbology/qgspointdisplacementrenderer.h:
```diff
@@ -99,6 +99,19 @@ class CORE_EXPORT QgsPointDisplacementRenderer: public QgsPointDistanceRenderer
+     * Sets a factor for increasing the label distances from the symbol.
+     * \param factor addition factor
+     * \see labelDistanceFactor()
+     */
```
\since QGIS 3.8

HenrikSpa replied:
I didn't see this comment before, but it's added now.

nyalldawson commented on src/core/symbology/qgspointdisplacementrenderer.h:
```diff
@@ -99,6 +99,19 @@ class CORE_EXPORT QgsPointDisplacementRenderer: public QgsPointDistanceRenderer
+    /**
+     * Returns the factor for label distance from the symbol.
+     * \see setLabelDistanceFactor()
+     */
```
\since QGIS 3.8

HenrikSpa replied:
I didn't see this comment before, but it's added now!

nyalldawson commented:
Looking good! Can you add some unit tests covering the new logic? Let me know if you need direction in how to do this...

nyalldawson committed:
Update src/core/symbology/qgspointdisplacementrenderer.cpp

Co-Authored-By: HenrikSpa <henrik.spangmyr@midvatten.se>

nyalldawson committed:
Update src/core/symbology/qgspointdisplacementrenderer.cpp

Co-Authored-By: HenrikSpa <henrik.spangmyr@midvatten.se>

Commit:
Changed QList to QVector for faster access.

HenrikSpa commented:
> Looking good! Can you add some unit tests covering the new logic? Let me know if you need direction in how to do this...

Great!
I've incorporated your suggestions, including QVector instead of QList.

For the unit tests however, I need some pointers.
I'm more used to python, so if the unit test could be done in tests/src/python/test_qgspointdisplacementrenderer.py it would be easier.
However, all affected members, the drawGroup and calculateSymbolAndLabelPositions are private members, so I guess they can't be accessed easily from the python interface? A unit test for calculateSymbolAndLabelPositions would be good.

So yes, pointers is needed if I'm to succeed in creating a unit test for the changes.

Commit:
Fixed indentation.

Commit:
Fixed indentation.

HenrikSpa commented:
Not stale

roya0045 commented:
Hi there,

I think you still need to add the /since tags in the function docs and adapt the sip files with that change.

As for the unit tests the variables such as the label shifts might be annoying to get. What you could do still by using python might be to generate an image and compare that with the desires results, this way this would help to show the proper behaviour of your change in the code.

For an existing test that behave this way, please take a look at 
[This for a generic image check example](https://github.com/qgis/QGIS/blob/411d642b864aa5a68d3d62a0d1d3ecf105b85c80/tests/src/python/test_qgslayoutlegend.py#L156) 
[This for a more useful example](https://github.com/qgis/QGIS/blob/411d642b864aa5a68d3d62a0d1d3ecf105b85c80/tests/src/python/test_qgspointdisplacementrenderer.py#L177), I am aware that you already fiddled with this test but [ this thing ](https://github.com/qgis/QGIS/blob/22772ea5c9f3876bc6b6ffbc34b8ec1d5bd42c79/src/core/qgsmultirenderchecker.cpp) would be the best way to ensure that your feature work properly. 

How I would go on about this is either to break a test with an image of your feature, copy the test to a new one, save the image generated in cdash for the failed code, fix the original test and use the generated image as the truth value for the test. another way would be do the test locally with python, to generate the ground truth image and copy what you did exactly to obtain the image for the new test.

Personnally I would suggest using existing test data in the test_data to perform the test but if none fit your need you can also add your own.

Hope this helps!

Commit:
Added tests for pointdisplacementlabels

HenrikSpa commented:
It helped!

I'm working on the tests.

Commit:
Added tests for PointDisplacementRenderer labels.

Commit:
Spelling

roya0045 commented:
[I'm not sure what might be causing the slight shift in label](https://dash.orfeo-toolbox.org/testDetails.php?test=104394874&build=357460), but the test look good in general, though there was a segfault. 

Would it be possible that a nullptr might be used somewhere in your code? I would advise to add a safety check on any input pointers to see if the pointer is valid to prevent segfault ( it took me time to fix some of the segfault in my PR..). Usually how this seems to be done is with an if statement in the first lines of the function.

I'm not sure if the label display can be fixed with a simple value or if it's the labelling engine being quirky, you cold try to grab the generated image to ensure that the output match the generated image since recreating the exact conditions of the testing instance might be hard.

Keep up the good work, this is a nice PR!

roya0045 commented:
In retrospective the segfault might be coming from the tests, where a null value is passed on somewhat rather than from your c++ modifications. But I may be wrong, maybe @nyalldawson would have a better idea as the segfault information seem to come from thread handling. 

A way to play it safer might be to not rely on class member of functions if possible since if those are modified during a test and are called in another test at the same time this might explain the segfault, but take what I say with a grain of salt.

HenrikSpa commented:
Thanks for the ideas!
I will run the test (locally) using the current master to see if it's my
addition that is the cause or if it was there before. The labels were not
tested before.

Den sön 24 mars 2019 23:09Alex <notifications@github.com> skrev:

> In retrospective the segfault might be coming from the tests, where a null
> value is passed on somewhat rather than from your c++ modifications. But I
> may be wrong, maybe @nyalldawson <https://github.com/nyalldawson> would
> have a better idea as the segfault information seem to come from thread
> handling.
>
> A way to play it safer might be to not rely on class member of functions
> if possible since if those are modified during a test and are called in
> another test at the same time this might explain the segfault, but take
> what I say with a grain of salt.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/9085#issuecomment-476005611>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQzJIbQa9Tru7zCCkPqPVhiBsAVKQaCtks5vZ_eagaJpZM4aguyd>
> .
>

HenrikSpa commented:
The same segmentation fault occurs when I run the label tests against master.

roya0045 commented:
I think the segfault occurs from the way the test are handled as I stated previously, could you try and declare everything in the function test itself and not use things such as self.

My hypothesis is that the travis tests may be done in parallel and when some tests reference the same functions, if one modifies the object that is pointer by the pointer in the c++ code, this will break the pointer and cause a segfault. 

I will try to look at it more in depth as soon as I can .

If you check [this test](https://github.com/qgis/QGIS/blob/1942507fe9a70363d15e027d260afed45c9680ce/tests/src/python/test_qgslayoutlegend.py#L156) they use and declare everything locally other than the assert. If you could try to declare everything locally and give it a try this just might solve the segfault.

roya0045 commented:
Yeah just rechecked the latest test, now the wait condition for the mutex caused the crash.

I'm not sure what might be causing this in relation to your code. But its not coming from the internals of what you changed but the vector handling and possibly the test somehow. So trying out the changes I suggested previously might solve this, or possibly not.

Commit:
Decoupled tests by removing shared self-variables.

Commit:
Fixed turned of testss

HenrikSpa commented:
I've updated the tests so that there are no shared objects, except for QgsProject.instance().

I still have shared functions to reduce duplicate code, but they are pure function calls with no side effects from what I can see except for QgsProject.instance(). Or perhaps QgsVectorLayer(myShpFile, ...) produces some side effect that I'm not aware of.

I still get segmentation fault when running the tests though.

roya0045 commented:
Ok, the shared variable might not be the issue, I made a gist with the previous version of the code and some changes if you wanted to try it out https://gist.github.com/roya0045/397f7379861097de8703322ffd2bbf6c

I doubt this will change anything but this is just my way to revert the changes that I suggested since I feel like I made you waste time and wanted to apologise.

I have thought about adding a 0.1 sleep time before evaluation if rendering was done in a thread and if the evaluation is done before rendering is done this might be the cause but that is still speculation on my part.

My other guess would be that  it might be something with the internals.

I'm just using the tried and true method of checking the simplest things first.

nyalldawson committed:
Fix segfault on exit

nyalldawson committed:
Test improvements

- Use standard test font, larger font size to minimize cross platform
rendering differences
- Create a local test report for debugging

nyalldawson commented:
@HenrikSpa The segfault on test exit is solved by https://github.com/qgis/QGIS/pull/9085/commits/253aad5c5aea743465b7d8a062ad55645d9d812c. Yep - it's a completely mystical fix! I only know this one because I've encountered that particular stack trace before. (Programming is fun! 🎈 )

I've also pushed a change here which improves the cross-platform stability of the new tests -- basically we always need to use a standard test font, and any tests using text needs to use big bold text. Otherwise the cross-platform text rendering differences makes the tests fragile.

nyalldawson committed:
Add test mask images

HenrikSpa commented:
Great! Thanks for all the help!

roya0045 commented:
@HenrikSpa Grats on the merge and thanks for the feature!
@nyalldawson Thanks for the help Nyall!

HenrikSpa commented:
Thanks! :D
